### PR TITLE
[WIP] feat: server port on server start up

### LINKED ISSUE #633: Return the server start on starting the server
**Is your feature request related to a problem? Please describe.**
When `0` is provided as the port value, the server will bind to any available free port. There is no way to get what port the server is using at the moment.

**Describe the solution you'd like**
I would like it if the `start` method on the server could return an `Int` representing the port the server is using.

**Describe alternatives you've considered**
There is no alternative way to get the port info.

**Additional context**
ServerBootstrap has an API to extract the port information.

### PR DESCRIPTION
**Pull request by girdharshubham**
closes #633

### CODE REVIEW

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 435091.37**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 431534.65**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 435096.51**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 440868.33**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 427372.25**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 436382.92**

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 1]
```diff
@@ -120,10 +121,11 @@ object Server {
       reqHandler      = settings.app.compile(zExec, settings, ServerTimeGenerator.make)
       init            = ServerChannelInitializer(zExec, settings, reqHandler)
       serverBootstrap = new ServerBootstrap().channelFactory(channelFactory).group(eventLoopGroup)
-      _ <- ChannelFuture.asManaged(serverBootstrap.childHandler(init).bind(settings.address))
-
+      channel <- ZManaged.effect(serverBootstrap.childHandler(init).bind(settings.address))
+      _       <- ChannelFuture.asManaged(channel)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]: why these changes?

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala [hunk 2]
```diff
@@ -16,7 +16,7 @@ import zio.{Has, ZIO, ZManaged}
 abstract class HttpRunnableSpec(port: Int) extends DefaultRunnableSpec { self =>
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]:
```suggestion
abstract class HttpRunnableSpec extends DefaultRunnableSpec { self =>
```

File: zio-http/src/main/scala/zhttp/service/Server.scala (hunk 1 above)
Comment (girdharshubham) [reply zio-http/src/main/scala/zhttp/service/Server.scala]: @tusharmath My bad. We don't really need this lift. Let me push a refactor.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 435173.76**

File: example/src/main/scala/example/HelloWorld.scala [hunk 3]
```diff
@@ -13,5 +13,5 @@ object HelloWorld extends App {
 
   // Run it like any simple app
   override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] =
-    Server.start(8090, app.silent).exitCode
+    Server.start(0, app.silent).exitCode
```
Comment (tusharmath) [inline example/src/main/scala/example/HelloWorld.scala]: this change isn't requried.

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala (hunk 2 above)
Comment (tusharmath) [reply zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]: We don't need to pass the port anymore for tests. Which would be great for tests.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 440279.20**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 440727.70**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 439637.11**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 425614.11**

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 4]
```diff
@@ -120,10 +121,11 @@ object Server {
       reqHandler      = settings.app.compile(zExec, settings, ServerTimeGenerator.make)
       init            = ServerChannelInitializer(zExec, settings, reqHandler)
       serverBootstrap = new ServerBootstrap().channelFactory(channelFactory).group(eventLoopGroup)
-      _ <- ChannelFuture.asManaged(serverBootstrap.childHandler(init).bind(settings.address))
-
+      channel         = serverBootstrap.childHandler(init).bind(settings.address)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala:124]: This is an effectful computation. Keep it as before, ie pass by name.

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala [hunk 5]
```diff
@@ -13,13 +13,14 @@ import zio.{Has, ZIO, ZManaged}
  * backend. For most of the other use cases directly running the HttpApp should suffice. HttpRunnableSpec spins of an
  * actual Http server and makes requests.
  */
-abstract class HttpRunnableSpec(port: Int) extends DefaultRunnableSpec { self =>
+abstract class HttpRunnableSpec extends DefaultRunnableSpec { self =>
+  type Port = Int
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]: why?

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala [hunk 6]
```diff
@@ -13,13 +13,14 @@ import zio.{Has, ZIO, ZManaged}
...
   def serve[R <: Has[_]](
     app: HttpApp[R, Throwable],
-  ): ZManaged[R with EventLoopGroup with ServerChannelFactory, Nothing, Unit] =
-    Server.make(Server.app(app) ++ Server.port(port) ++ Server.paranoidLeakDetection).orDie
+  ): ZManaged[R with EventLoopGroup with ServerChannelFactory, Nothing, Int] =
+    Server.make(Server.app(app) ++ Server.port(0) ++ Server.paranoidLeakDetection).orDie
 
-  def status(path: Path): ZIO[EventLoopGroup with ChannelFactory, Throwable, Status] =
+  def status(path: Path)(implicit port: Port): ZIO[EventLoopGroup with ChannelFactory, Throwable, Status] =
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]: don't use implicits for data.

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 7]
```diff
@@ -136,7 +136,7 @@ object ServerSpec extends HttpRunnableSpec(8088) {
     }
   }
 
-  def staticAppSpec = suite("StaticAppSpec") {
+  def staticAppSpec(implicit port: Port) = suite("StaticAppSpec") {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: drop all implicit ports.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 443575.60**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 440463.11**

File: zio-http/src/main/scala/zhttp/service/Server.scala (hunk 4 above)
Comment (girdharshubham) [reply zio-http/src/main/scala/zhttp/service/Server.scala:124]: Won't the effect be run later on during the useForever call? and since it's in ZManaged, acquire and release would be handled?

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 434346.30**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 430472.77**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 434653.64**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 438450.22**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 443847.30**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 444502.16**

File: zio-http/src/main/scala/zhttp/service/Server.scala (hunk 4 above)
Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/service/Server.scala:124]: It's not the same. The previous approach can be caught and is a checked exception.
Here however, it will be thrown as an unchecked exception and you will not be able to catch it.

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala [hunk 8]
```diff
@@ -13,10 +13,10 @@ import zio.{Has, ZIO, ZManaged}
  * backend. For most of the other use cases directly running the HttpApp should suffice. HttpRunnableSpec spins of an
  * actual Http server and makes requests.
  */
-abstract class HttpRunnableSpec(port: Int) extends DefaultRunnableSpec { self =>
+abstract class HttpRunnableSpec(port: Int = 8088) extends DefaultRunnableSpec { self =>
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]: I think port need not be specified. Automatically a random port should be assigned for the tests.

Comment (tusharmath) [reply zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]: OR — set default port value to `0`

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 436320.05**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 439277.89**

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala (hunk 8 above)
Comment (girdharshubham) [reply zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]: Done.
Thanks for the suggestion! @tusharmath

**Comment by girdharshubham** [comment]
Taking down this PR because of it being a wip
